prov/verbs/src/verbs_info.c

```diff
@@ -1007,6 +1012,162 @@ static int vrb_ifa_rdma_info(const struct ifaddrs *ifa, char **dev_name,
 	return ret;
 }
 
+int vrb_get_port_space(const struct fi_info *info)
+{
+	if (info != NULL && info->addr_format == FI_SOCKADDR_IB)
+		return RDMA_PS_IB;
+	else
+		return RDMA_PS_TCP;
+}
+
+static int vrb_gid_is_null(const union ibv_gid *gid)
+{
+	return !(gid->raw[8]  | gid->raw[9]  | gid->raw[10] | gid->raw[11] |
+	         gid->raw[12] | gid->raw[13] | gid->raw[14] | gid->raw[15]);
+}
+
+static int vrb_gid_to_str(const union ibv_gid *gid, char gid_str[40])
+{
+	if (sprintf(gid_str, "%02x%02x:%02x%02x:%02x%02x:%02x%02x:%02x%02x:%02x%02x:%02x%02x:%02x%02x",
+		        gid->raw[ 0], gid->raw[ 1],
+		        gid->raw[ 2], gid->raw[ 3],
+		        gid->raw[ 4], gid->raw[ 5],
+		        gid->raw[ 6], gid->raw[ 7],
+		        gid->raw[ 8], gid->raw[ 9],
+		        gid->raw[10], gid->raw[11],
+		        gid->raw[12], gid->raw[13],
+		        gid->raw[14], gid->raw[15]) != 39)
+	{
+		return -FI_EINVAL;
+	}
+	return 0;
+}
+
+static int vrb_gid_rdma_info(char gid_str[40], char **dev_name,
+                             struct rdma_addrinfo **rai)
+{
+	struct rdma_cm_id *id = NULL;
+	struct rdma_addrinfo rai_hints = {
+		.ai_flags = RAI_PASSIVE | RAI_NUMERICHOST | RAI_FAMILY,
+		.ai_family = AF_IB,
+		.ai_port_space = RDMA_PS_IB,
+	}, *rai_ = NULL;
+	int ret;
+
+	ret = rdma_create_id(NULL, &id, NULL, rai_hints.ai_port_space);
+	if (ret) {
+		ret = -errno;
+		goto err;
+	}
+
+	ret = rdma_getaddrinfo(gid_str, NULL, &rai_hints, &rai_);
+	if (ret) {
+		ret = -errno;
+		goto err;
+	}
+
+	ret = rdma_bind_addr(id, rai_->ai_src_addr);
+	if (ret) {
+		ret = -errno;
+		goto err;
+	}
+
+	if (!id->verbs) {
+		ret = -FI_EINVAL;
+		goto err;
+	}
+
+	*dev_name = strdup(ibv_get_device_name(id->verbs->device));
+	if (!(!dev_name)) {
+		ret = -FI_ENOMEM;
+		goto err;
+	}
+
+	rdma_destroy_id(id);
+	*rai = rai_;
+	return 0;
+
+err:
+	if (rai_)
+		rdma_freeaddrinfo(rai_);
+	if (id)
+		rdma_destroy_id(id);
+	return ret;
+}
+
+static int vrb_get_gids(struct dlist_entry *verbs_devs)
+{
+	struct rdma_addrinfo *rai = NULL;
+	struct ibv_device **devices;
+	char *dev_name = NULL;
+	int num_devices;
+	struct ibv_context *context;
+	int ret, num_verbs_ifs = 0;
+	struct ibv_device_attr device_attr;
+	struct ibv_port_attr port_attr;
+	union ibv_gid gid;
+	char gid_str[40];
+
+	devices = ibv_get_device_list(&num_devices);
+	if (!devices)
+		return -errno;
+
+	for (int dev = 0; dev < num_devices; dev++) {
+		context = ibv_open_device(devices[dev]);
+
+		ret = ibv_query_device(context, &device_attr);
```

Comment:
GID addressing should be limited to devices using the IB link layer.  I don't believe iWarp or RoCE will work.

Opening and querying devices is an expensive operation.  The verbs code does this already somewhere.  Obtaining the GIDs should be moved there to avoid duplicating these costs.  (That said, holy cow, is the verbs initialization code complex.)

Reply:
Thanks. I see that `vrb_have_device()` opens the devices. I'll merge the two functions together.

Initially, I wanted to use the function `rdma_getaddrinfo()` to retrieve a complete list of local RDMA devices. However, during my testing, the function was always returning only 1 single device and the pointer `ai_next` to the next device was always NULL. Not sure if it's a bug in RDMACM or because I'm using `rdma_getaddrinfo()` incorrectly. 
That's the reason why I had to deal with all this verbs initialization code :-/.